### PR TITLE
Make sure EventPipeProvider::m_keywords is accurate after EventPipeConfiguration::Disable

### LINKED ISSUE #25298: EventPipe keywords/level should be re-evaluated when a session is disabled
This issue is tracking:

https://github.com/dotnet/coreclr/issues/25069#issuecomment-501258814

### PR DESCRIPTION
**Pull request by cshung**
Fixes #25298 

**(1st commit)**

1. Added a method `EventPipeConfiguration::ComputeKeywordAndLevel` to compute the bitwise or of all the keywords and the maximum level across all sessions.
2. Call `EventPipeConfiguration::ComputeKeywordAndLevel` everywhere `EventPipeProvider::SetConfiguration` or `EventPipeProvider::UnsetConfiguration` is called, pass the combined value into it instead of having it to compute itself.
3. Removed the logic to compute union and maximum inside `EventPipeProvider::SetConfiguration`, the computation is now done in `EventPipeConfiguration::ComputeKeywordAndLevel`.
4. Passing the combined keyword and level to `RefreshAllEvents` instead of the per session keyword and level there, I believe that was a bug.
5. Remove the unnecessary `sessionId` parameter to `RefreshAllEvents`
6. Call `RefreshAllEvents` on `EventPipeProvider::UnsetConfiguration` since the combined keyword and level could change at that point.

**(2nd commit)**
1. Adding const everywhere I needed to allow the wanted const constraint on `EventPipeConfiguration::ComputeKeywordAndLevel`

**(3rd commit)**
1. CI discovered a bug in my previous change, if the session is removed from `EventPipe::s_pSessions` before it is re-enabled for rundown, `ComputeKeywordAndLevel` would be wrong because the rundown provider would think there is no session listening to it while that's not true.

**(4th commit)**
1. Removed all other unused parameters to `RefreshAllEvents` (Thanks Sung for the comment)
2. Moved the computation of keyword and level so that we save the computation time when it is unused. (Thanks Noah for the comment)

### CODE REVIEW

File: src/vm/eventpipeprovider.h [hunk 1]
```diff
@@ -97,14 +99,15 @@ class EventPipeProvider
...
+        INT64 keywordsForAllSessions,
+        EventPipeEventLevel providerLevelForAllSessions,
         uint64_t sessionId,
         INT64 keywords,
         EventPipeEventLevel providerLevel,
         LPCWSTR pFilterData);
 
     // Refresh the runtime state of all events.
     void RefreshAllEvents(
-        uint64_t sessionId,
```
Comment (sywhang) [inline src/vm/eventpipeprovider.h]: Synced with you offline - We should remove the other two arguments since they're not used either.

**Comment by jorive** [comment]
@cshung Thanks for using `const` where needed :+1:

**Comment by jorive** [comment]
This PR restores the global state of providers/keywords/level after a session has been disabled. Have we tested these changes with user defined providers?

File: src/vm/eventpipeprovider.h (hunk 1 above)
Comment (jorive) [reply src/vm/eventpipeprovider.h]: Agreed, this is a left over ... I would have expected a warning treated as error for unused variables/params :(

File: src/vm/eventpipeprovider.cpp [hunk 2]
```diff
@@ -152,6 +153,11 @@ EventPipeProviderCallbackData EventPipeProvider::UnsetConfiguration(
 
     if (m_sessions & sessionMask)
         m_sessions &= ~sessionMask;
+
+    m_keywords = keywordsForAllSessions;
+    m_providerLevel = providerLevelForAllSessions;
+
+    RefreshAllEvents(m_keywords, m_providerLevel);
```
Comment (noahfalk) [inline src/vm/eventpipeprovider.cpp]: To ensure no confusion I commented on #25232 that this change partially mitigates that issue but does not fully solve it. The title and description of this PR are entirely accurate, I'm just heading off the potential for it to be misconstrued.

File: src/vm/eventpipeconfiguration.cpp [hunk 3]
```diff
@@ -335,12 +371,18 @@ void EventPipeConfiguration::Disable(const EventPipeSession &session, EventPipeP
         {
             EventPipeProvider *pProvider = pElem->GetValue();
 
+            INT64 keywordForAllSessions;
```
[the remark this answers is not in the document]
Comment (cshung) [reply src/vm/eventpipeconfiguration.cpp]: Will be addressed in the next iteration.

File: src/vm/eventpipeconfiguration.cpp [hunk 4]
```diff
@@ -298,12 +328,18 @@ void EventPipeConfiguration::Enable(EventPipeSession &session, EventPipeProvider
...
```
[the remark this answers is not in the document]
Comment (cshung) [reply src/vm/eventpipeconfiguration.cpp]: Will be addressed in the next iteration.

File: src/vm/eventpipeprovider.h (hunk 1 above)
Comment (cshung) [reply src/vm/eventpipeprovider.h]: Will be addressed in the next iteration.

File: src/vm/eventpipe.cpp [hunk 5]
```diff
@@ -339,13 +339,6 @@ void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallback
     // Disable pSession tracing.
     s_config.Disable(*pSession, pEventPipeProviderCallbackDataQueue);
 
-    // At this point, we should not be writing events to this session anymore
-    // This is a good time to remove the session from the array.
-    _ASSERTE(s_pSessions[pSession->GetIndex()] == pSession);
-
-    // Remove the session from the array, and mask.
-    s_pSessions[pSession->GetIndex()].Store(nullptr);
```
[the remark this answers is not in the document]
Comment (cshung) [reply src/vm/eventpipe.cpp]: The issue is tracked as https://github.com/dotnet/coreclr/issues/25430.

**Pull request merged**